In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.DataFrame({
    'salary':np.random.randint(30, 90, 20),
    'height':np.random.randint(120, 190, 20),
    'weight':np.random.randint(40, 90, 20),
    'age': np.random.randint(20, 35, 20),
    'country':np.random.choice(['india', 'usa', 'aus', 'china'], 20),
    'gender': np.random.choice(['male', 'female'], 20)
}, index=[chr(i) for i in range(65, 65+20)])
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,male
B,79,152,80,20,usa,male
C,33,142,47,25,india,female
D,44,162,83,21,usa,male
E,77,143,80,23,aus,female
F,33,157,74,31,usa,female
G,70,162,60,33,aus,female
H,75,168,47,24,usa,female
I,59,164,40,32,india,female
J,33,124,67,22,india,female


     apply

In [3]:
df['height']

A    170
B    152
C    142
D    162
E    143
F    157
G    162
H    168
I    164
J    124
K    189
L    126
M    180
N    141
O    180
P    151
Q    155
R    189
S    133
T    172
Name: height, dtype: int32

In [6]:
df['height'].apply(lambda x: f"{x} cm")


A    170 cm
B    152 cm
C    142 cm
D    162 cm
E    143 cm
F    157 cm
G    162 cm
H    168 cm
I    164 cm
J    124 cm
K    189 cm
L    126 cm
M    180 cm
N    141 cm
O    180 cm
P    151 cm
Q    155 cm
R    189 cm
S    133 cm
T    172 cm
Name: height, dtype: object

In [7]:
df['height'] = df['height'].apply(lambda x: f"{x} cm")


In [8]:
df

,salary,height,weight,age,country,gender
A,60,170 cm,52,31,aus,male
B,79,152 cm,80,20,usa,male
C,33,142 cm,47,25,india,female
D,44,162 cm,83,21,usa,male
E,77,143 cm,80,23,aus,female
F,33,157 cm,74,31,usa,female
G,70,162 cm,60,33,aus,female
H,75,168 cm,47,24,usa,female
I,59,164 cm,40,32,india,female
J,33,124 cm,67,22,india,female


In [5]:
for i in df['height']:
    print(f"{i} cm")

170 cm
152 cm
142 cm
162 cm
143 cm
157 cm
162 cm
168 cm
164 cm
124 cm
189 cm
126 cm
180 cm
141 cm
180 cm
151 cm
155 cm
189 cm
133 cm
172 cm


In [9]:
df['height']

A    170 cm
B    152 cm
C    142 cm
D    162 cm
E    143 cm
F    157 cm
G    162 cm
H    168 cm
I    164 cm
J    124 cm
K    189 cm
L    126 cm
M    180 cm
N    141 cm
O    180 cm
P    151 cm
Q    155 cm
R    189 cm
S    133 cm
T    172 cm
Name: height, dtype: object

In [10]:
df

,salary,height,weight,age,country,gender
A,60,170 cm,52,31,aus,male
B,79,152 cm,80,20,usa,male
C,33,142 cm,47,25,india,female
D,44,162 cm,83,21,usa,male
E,77,143 cm,80,23,aus,female
F,33,157 cm,74,31,usa,female
G,70,162 cm,60,33,aus,female
H,75,168 cm,47,24,usa,female
I,59,164 cm,40,32,india,female
J,33,124 cm,67,22,india,female


In [14]:
df['height'] = df['height'].apply(lambda x: int(x.split()[0]))

In [15]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,male
B,79,152,80,20,usa,male
C,33,142,47,25,india,female
D,44,162,83,21,usa,male
E,77,143,80,23,aus,female
F,33,157,74,31,usa,female
G,70,162,60,33,aus,female
H,75,168,47,24,usa,female
I,59,164,40,32,india,female
J,33,124,67,22,india,female


In [17]:
df['gender'].apply(lambda x: x.upper())

A      MALE
B      MALE
C    FEMALE
D      MALE
E    FEMALE
F    FEMALE
G    FEMALE
H    FEMALE
I    FEMALE
J    FEMALE
K      MALE
L      MALE
M    FEMALE
N      MALE
O      MALE
P    FEMALE
Q    FEMALE
R      MALE
S    FEMALE
T      MALE
Name: gender, dtype: object

In [ ]:
def func(x):
    return x.upper()

In [18]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,male
B,79,152,80,20,usa,male
C,33,142,47,25,india,female
D,44,162,83,21,usa,male
E,77,143,80,23,aus,female
F,33,157,74,31,usa,female
G,70,162,60,33,aus,female
H,75,168,47,24,usa,female
I,59,164,40,32,india,female
J,33,124,67,22,india,female


In [20]:
df['gender'] = df['gender'].map({'male':'M', 'female':'F'})

In [21]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [23]:
df[df['country'] == 'india']

,salary,height,weight,age,country,gender
C,33,142,47,25,india,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F
N,68,141,73,29,india,M
S,54,133,63,22,india,F


In [24]:
df[(df['country'] == 'india') & (df['gender'] == 'M')]


,salary,height,weight,age,country,gender
N,68,141,73,29,india,M


In [25]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [30]:
df[df['country'] == 'india']['salary'].mean()

49.4

In [33]:
df[(df['country'] == 'usa') & (df['gender'] == 'M')]['salary'].mean()

68.5

In [34]:
df[(df['country'] == 'usa') & (df['gender'] == 'F')]['salary'].mean()


53.333333333333336

In [35]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [36]:
df.groupby('country')

In [37]:
df.groupby('country').groups


{'aus': ['A', 'E', 'G', 'T'], 'china': ['K', 'M', 'O', 'P'], 'india': ['C', 'I', 'J', 'N', 'S'], 'usa': ['B', 'D', 'F', 'H', 'L', 'Q', 'R']}

In [38]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [39]:
df.groupby('country')['salary'].mean()


country
aus      60.50
china    58.25
india    49.40
usa      62.00
Name: salary, dtype: float64

In [40]:
df.groupby('country')[['salary', 'weight', 'height']].mean()


,salary,weight,height
country,,,
aus,60.50,63.500000,161.750000
china,58.25,81.000000,175.000000
india,49.40,58.000000,140.800000
usa,62.00,67.857143,158.428571


In [41]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [42]:
df.groupby('country').agg({
    'salary':'mean',
    'weight':['mean', 'median']
})


salary     weight       
          mean       mean median
country                         
aus      60.50  63.500000   61.0
china    58.25  81.000000   88.5
india    49.40  58.000000   63.0
usa      62.00  67.857143   74.0

In [43]:
df.groupby(['country', 'gender'])['salary'].mean()

country  gender
aus      F         73.500000
         M         47.500000
china    F         61.500000
         M         55.000000
india    F         44.750000
         M         68.000000
usa      F         53.333333
         M         68.500000
Name: salary, dtype: float64

In [44]:
df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
})


salary height    age       
                     mean median   mean median
country gender                                
aus     F       73.500000  152.5  28.00   28.0
        M       47.500000  171.0  28.50   28.5
china   F       61.500000  165.5  33.50   33.5
        M       55.000000  184.5  21.50   21.5
india   F       44.750000  137.5  25.25   23.5
        M       68.000000  141.0  29.00   29.0
usa     F       53.333333  157.0  25.00   24.0
        M       68.500000  157.0  23.50   22.5

In [45]:
df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
    'gender':'count'
})


salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [46]:
df.groupby(['country'])

In [48]:
idx = df.groupby(['country']).groups
idx

{'aus': ['A', 'E', 'G', 'T'], 'china': ['K', 'M', 'O', 'P'], 'india': ['C', 'I', 'J', 'N', 'S'], 'usa': ['B', 'D', 'F', 'H', 'L', 'Q', 'R']}

In [49]:
idx['aus']

Index(['A', 'E', 'G', 'T'], dtype='object')

In [50]:
df.loc[idx['aus']]

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
E,77,143,80,23,aus,F
G,70,162,60,33,aus,F
T,35,172,62,26,aus,M


In [51]:
df.loc[idx['india']]


,salary,height,weight,age,country,gender
C,33,142,47,25,india,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F
N,68,141,73,29,india,M
S,54,133,63,22,india,F


In [52]:
df1 = df.groupby(['country', 'gender']).agg({
    'salary':'mean',
    'height':'median',
    'age':['mean', 'median'],
    'gender':'count'
})
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [53]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [54]:
df.index

Index(['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N',
       'O', 'P', 'Q', 'R', 'S', 'T'],
      dtype='object')

In [55]:
df.columns

Index(['salary', 'height', 'weight', 'age', 'country', 'gender'], dtype='object')

In [56]:
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [57]:
df1.index

MultiIndex([(  'aus', 'F'),
            (  'aus', 'M'),
            ('china', 'F'),
            ('china', 'M'),
            ('india', 'F'),
            ('india', 'M'),
            (  'usa', 'F'),
            (  'usa', 'M')],
           names=['country', 'gender'])

In [58]:
print(dir(pd))

['ArrowDtype', 'BooleanDtype', 'Categorical', 'CategoricalDtype', 'CategoricalIndex', 'DataFrame', 'DateOffset', 'DatetimeIndex', 'DatetimeTZDtype', 'ExcelFile', 'ExcelWriter', 'Flags', 'Float32Dtype', 'Float64Dtype', 'Grouper', 'HDFStore', 'Index', 'IndexSlice', 'Int16Dtype', 'Int32Dtype', 'Int64Dtype', 'Int8Dtype', 'Interval', 'IntervalDtype', 'IntervalIndex', 'MultiIndex', 'NA', 'NaT', 'NamedAgg', 'Period', 'PeriodDtype', 'PeriodIndex', 'RangeIndex', 'Series', 'SparseDtype', 'StringDtype', 'Timedelta', 'TimedeltaIndex', 'Timestamp', 'UInt16Dtype', 'UInt32Dtype', 'UInt64Dtype', 'UInt8Dtype', '__all__', '__builtins__', '__cached__', '__doc__', '__docformat__', '__file__', '__git_version__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_built_with_meson', '_config', '_is_numpy_dev', '_libs', '_pandas_datetime_CAPI', '_pandas_parser_CAPI', '_testing', '_typing', '_version_meson', 'annotations', 'api', 'array', 'arrays', 'bdate_range', 'compat', 'conca

In [59]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [60]:
print(dir(df))

['T', '_AXIS_LEN', '_AXIS_ORDERS', '_AXIS_TO_AXIS_NUMBER', '_HANDLED_TYPES', '__abs__', '__add__', '__and__', '__annotations__', '__array__', '__array_priority__', '__array_ufunc__', '__bool__', '__class__', '__contains__', '__copy__', '__dataframe__', '__dataframe_consortium_standard__', '__deepcopy__', '__delattr__', '__delitem__', '__dict__', '__dir__', '__divmod__', '__doc__', '__eq__', '__finalize__', '__floordiv__', '__format__', '__ge__', '__getattr__', '__getattribute__', '__getitem__', '__getstate__', '__gt__', '__hash__', '__iadd__', '__iand__', '__ifloordiv__', '__imod__', '__imul__', '__init__', '__init_subclass__', '__invert__', '__ior__', '__ipow__', '__isub__', '__iter__', '__itruediv__', '__ixor__', '__le__', '__len__', '__lt__', '__matmul__', '__mod__', '__module__', '__mul__', '__ne__', '__neg__', '__new__', '__nonzero__', '__or__', '__pandas_priority__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloord

In [61]:
len(dir(df))


444

In [62]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [64]:
print(df.to_dict())

{'salary': {'A': 60, 'B': 79, 'C': 33, 'D': 44, 'E': 77, 'F': 33, 'G': 70, 'H': 75, 'I': 59, 'J': 33, 'K': 32, 'L': 81, 'M': 64, 'N': 68, 'O': 78, 'P': 59, 'Q': 52, 'R': 70, 'S': 54, 'T': 35}, 'height': {'A': 170, 'B': 152, 'C': 142, 'D': 162, 'E': 143, 'F': 157, 'G': 162, 'H': 168, 'I': 164, 'J': 124, 'K': 189, 'L': 126, 'M': 180, 'N': 141, 'O': 180, 'P': 151, 'Q': 155, 'R': 189, 'S': 133, 'T': 172}, 'weight': {'A': 52, 'B': 80, 'C': 47, 'D': 83, 'E': 80, 'F': 74, 'G': 60, 'H': 47, 'I': 40, 'J': 67, 'K': 89, 'L': 77, 'M': 89, 'N': 73, 'O': 58, 'P': 88, 'Q': 47, 'R': 67, 'S': 63, 'T': 62}, 'age': {'A': 31, 'B': 20, 'C': 25, 'D': 21, 'E': 23, 'F': 31, 'G': 33, 'H': 24, 'I': 32, 'J': 22, 'K': 20, 'L': 29, 'M': 33, 'N': 29, 'O': 23, 'P': 34, 'Q': 20, 'R': 24, 'S': 22, 'T': 26}, 'country': {'A': 'aus', 'B': 'usa', 'C': 'india', 'D': 'usa', 'E': 'aus', 'F': 'usa', 'G': 'aus', 'H': 'usa', 'I': 'india', 'J': 'india', 'K': 'china', 'L': 'usa', 'M': 'china', 'N': 'india', 'O': 'china', 'P': 'ch

In [65]:
df.to_html()

'<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>salary</th>\n      <th>height</th>\n      <th>weight</th>\n      <th>age</th>\n      <th>country</th>\n      <th>gender</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>A</th>\n      <td>60</td>\n      <td>170</td>\n      <td>52</td>\n      <td>31</td>\n      <td>aus</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>B</th>\n      <td>79</td>\n      <td>152</td>\n      <td>80</td>\n      <td>20</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>C</th>\n      <td>33</td>\n      <td>142</td>\n      <td>47</td>\n      <td>25</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>D</th>\n      <td>44</td>\n      <td>162</td>\n      <td>83</td>\n      <td>21</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>E</th>\n      <td>77</td>\n      <td>143</td>\n      <td>80</td>\n      <td>23</td>\n   

<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>salary</th>\n      <th>height</th>\n      <th>weight</th>\n      <th>age</th>\n      <th>country</th>\n      <th>gender</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>A</th>\n      <td>60</td>\n      <td>170</td>\n      <td>52</td>\n      <td>31</td>\n      <td>aus</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>B</th>\n      <td>79</td>\n      <td>152</td>\n      <td>80</td>\n      <td>20</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>C</th>\n      <td>33</td>\n      <td>142</td>\n      <td>47</td>\n      <td>25</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>D</th>\n      <td>44</td>\n      <td>162</td>\n      <td>83</td>\n      <td>21</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>E</th>\n      <td>77</td>\n      <td>143</td>\n      <td>80</td>\n      <td>23</td>\n      <td>aus</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>F</th>\n      <td>33</td>\n      <td>157</td>\n      <td>74</td>\n      <td>31</td>\n      <td>usa</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>G</th>\n      <td>70</td>\n      <td>162</td>\n      <td>60</td>\n      <td>33</td>\n      <td>aus</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>H</th>\n      <td>75</td>\n      <td>168</td>\n      <td>47</td>\n      <td>24</td>\n      <td>usa</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>I</th>\n      <td>59</td>\n      <td>164</td>\n      <td>40</td>\n      <td>32</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>J</th>\n      <td>33</td>\n      <td>124</td>\n      <td>67</td>\n      <td>22</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>K</th>\n      <td>32</td>\n      <td>189</td>\n      <td>89</td>\n      <td>20</td>\n      <td>china</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>L</th>\n      <td>81</td>\n      <td>126</td>\n      <td>77</td>\n      <td>29</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>M</th>\n      <td>64</td>\n      <td>180</td>\n      <td>89</td>\n      <td>33</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>N</th>\n      <td>68</td>\n      <td>141</td>\n      <td>73</td>\n      <td>29</td>\n      <td>india</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>O</th>\n      <td>78</td>\n      <td>180</td>\n      <td>58</td>\n      <td>23</td>\n      <td>china</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>P</th>\n      <td>59</td>\n      <td>151</td>\n      <td>88</td>\n      <td>34</td>\n      <td>china</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>Q</th>\n      <td>52</td>\n      <td>155</td>\n      <td>47</td>\n      <td>20</td>\n      <td>usa</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>R</th>\n      <td>70</td>\n      <td>189</td>\n      <td>67</td>\n      <td>24</td>\n      <td>usa</td>\n      <td>M</td>\n    </tr>\n    <tr>\n      <th>S</th>\n      <td>54</td>\n      <td>133</td>\n      <td>63</td>\n      <td>22</td>\n      <td>india</td>\n      <td>F</td>\n    </tr>\n    <tr>\n      <th>T</th>\n      <td>35</td>\n      <td>172</td>\n      <td>62</td>\n      <td>26</td>\n      <td>aus</td>\n      <td>M</td>\n    </tr>\n  </tbody>\n</table>'

In [66]:
df.to_numpy()

array([[60, 170, 52, 31, 'aus', 'M'],
       [79, 152, 80, 20, 'usa', 'M'],
       [33, 142, 47, 25, 'india', 'F'],
       [44, 162, 83, 21, 'usa', 'M'],
       [77, 143, 80, 23, 'aus', 'F'],
       [33, 157, 74, 31, 'usa', 'F'],
       [70, 162, 60, 33, 'aus', 'F'],
       [75, 168, 47, 24, 'usa', 'F'],
       [59, 164, 40, 32, 'india', 'F'],
       [33, 124, 67, 22, 'india', 'F'],
       [32, 189, 89, 20, 'china', 'M'],
       [81, 126, 77, 29, 'usa', 'M'],
       [64, 180, 89, 33, 'china', 'F'],
       [68, 141, 73, 29, 'india', 'M'],
       [78, 180, 58, 23, 'china', 'M'],
       [59, 151, 88, 34, 'china', 'F'],
       [52, 155, 47, 20, 'usa', 'F'],
       [70, 189, 67, 24, 'usa', 'M'],
       [54, 133, 63, 22, 'india', 'F'],
       [35, 172, 62, 26, 'aus', 'M']], dtype=object)

In [67]:
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [68]:
df

,salary,height,weight,age,country,gender
A,60,170,52,31,aus,M
B,79,152,80,20,usa,M
C,33,142,47,25,india,F
D,44,162,83,21,usa,M
E,77,143,80,23,aus,F
F,33,157,74,31,usa,F
G,70,162,60,33,aus,F
H,75,168,47,24,usa,F
I,59,164,40,32,india,F
J,33,124,67,22,india,F


In [70]:
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [71]:
df1.stack()

salary  height    age  gender
country gender                                         
aus     F      count         NaN     NaN    NaN     2.0
               mean    73.500000     NaN  28.00     NaN
               median        NaN   152.5  28.00     NaN
        M      count         NaN     NaN    NaN     2.0
               mean    47.500000     NaN  28.50     NaN
               median        NaN   171.0  28.50     NaN
china   F      count         NaN     NaN    NaN     2.0
               mean    61.500000     NaN  33.50     NaN
               median        NaN   165.5  33.50     NaN
        M      count         NaN     NaN    NaN     2.0
               mean    55.000000     NaN  21.50     NaN
               median        NaN   184.5  21.50     NaN
india   F      count         NaN     NaN    NaN     4.0
               mean    44.750000     NaN  25.25     NaN
               median        NaN   137.5  23.50     NaN
        M      count         NaN     NaN    NaN     1.0
               mean    68.000000     NaN  29.00     NaN
               median        NaN   141.0  29.00     NaN
usa     F      count         NaN     NaN    NaN     3.0
               mean    53.333333     NaN  25.00     NaN
               median        NaN   157.0  24.00     NaN
        M      count         NaN     NaN    NaN     4.0
               mean    68.500000     NaN  23.50     NaN
               median        NaN   157.0  22.50     NaN

In [73]:
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [74]:
df1.stack(level=0)

mean  median  count
country gender                                 
aus     F      age     28.000000    28.0    NaN
               gender        NaN     NaN    2.0
               height        NaN   152.5    NaN
               salary  73.500000     NaN    NaN
        M      age     28.500000    28.5    NaN
               gender        NaN     NaN    2.0
               height        NaN   171.0    NaN
               salary  47.500000     NaN    NaN
china   F      age     33.500000    33.5    NaN
               gender        NaN     NaN    2.0
               height        NaN   165.5    NaN
               salary  61.500000     NaN    NaN
        M      age     21.500000    21.5    NaN
               gender        NaN     NaN    2.0
               height        NaN   184.5    NaN
               salary  55.000000     NaN    NaN
india   F      age     25.250000    23.5    NaN
               gender        NaN     NaN    4.0
               height        NaN   137.5    NaN
               salary  44.750000     NaN    NaN
        M      age     29.000000    29.0    NaN
               gender        NaN     NaN    1.0
               height        NaN   141.0    NaN
               salary  68.000000     NaN    NaN
usa     F      age     25.000000    24.0    NaN
               gender        NaN     NaN    3.0
               height        NaN   157.0    NaN
               salary  53.333333     NaN    NaN
        M      age     23.500000    22.5    NaN
               gender        NaN     NaN    4.0
               height        NaN   157.0    NaN
               salary  68.500000     NaN    NaN

In [75]:
df1

salary height    age        gender
                     mean median   mean median  count
country gender                                       
aus     F       73.500000  152.5  28.00   28.0      2
        M       47.500000  171.0  28.50   28.5      2
china   F       61.500000  165.5  33.50   33.5      2
        M       55.000000  184.5  21.50   21.5      2
india   F       44.750000  137.5  25.25   23.5      4
        M       68.000000  141.0  29.00   29.0      1
usa     F       53.333333  157.0  25.00   24.0      3
        M       68.500000  157.0  23.50   22.5      4

In [76]:
df1.unstack()

salary       height           age                    gender   
              mean       median          mean       median        count   
gender           F     M      F      M      F     M      F     M      F  M
country                                                                   
aus      73.500000  47.5  152.5  171.0  28.00  28.5   28.0  28.5      2  2
china    61.500000  55.0  165.5  184.5  33.50  21.5   33.5  21.5      2  2
india    44.750000  68.0  137.5  141.0  25.25  29.0   23.5  29.0      4  1
usa      53.333333  68.5  157.0  157.0  25.00  23.5   24.0  22.5      3  4

In [79]:
df1.unstack().unstack()


                gender  country
salary  mean    F       aus         73.500000
                        china       61.500000
                        india       44.750000
                        usa         53.333333
                M       aus         47.500000
                        china       55.000000
                        india       68.000000
                        usa         68.500000
height  median  F       aus        152.500000
                        china      165.500000
                        india      137.500000
                        usa        157.000000
                M       aus        171.000000
                        china      184.500000
                        india      141.000000
                        usa        157.000000
age     mean    F       aus         28.000000
                        china       33.500000
                        india       25.250000
                        usa         25.000000
                M       aus         28.500000
  

In [81]:
data = pd.read_csv('data.csv')
data

,Unnamed: 0,Laptop,Ratings,prices
0,0,"ASUS Vivobook 16,Smartchoice, 13th Gen, Intel ...",4.2,"63,990"
1,1,"HP 15, AMD Ryzen 5 7520U,(16GB LPDDR5, 512GB S...",4.0,"54,690"
2,2,Lenovo V15 G4 AMD Athlon Silver 7120U Laptop 8...,4.0,"43,999"
3,3,More results,3.8,"59,990"
4,4,"Dell 15, 13th Gen Intel Core i5-1334U (16GB DD...",3.3,"39,990"
5,5,"Acer Smartchoice Aspire One, Intel Core Celero...",4.0,"50,490"
6,6,"Dell 15 (Previously Inspiron) Laptop, 14th Gen...",4.2,"72,120"
7,7,"acer Aspire 7, Intel Core 5-210H Processor, NV...",4.8,"49,480"
8,8,"HP 15, 13th Gen Intel Core i3-1315U Laptop (8G...",4.1,"68,990"
9,9,"ASUS TUF A15 (2025), AMD Ryzen 7 7445HS,RTX 30...",3.4,"63,990"


In [83]:
data.drop('Unnamed: 0', axis=1, inplace=True)

In [84]:
data

,Laptop,Ratings,prices
0,"ASUS Vivobook 16,Smartchoice, 13th Gen, Intel ...",4.2,"63,990"
1,"HP 15, AMD Ryzen 5 7520U,(16GB LPDDR5, 512GB S...",4.0,"54,690"
2,Lenovo V15 G4 AMD Athlon Silver 7120U Laptop 8...,4.0,"43,999"
3,More results,3.8,"59,990"
4,"Dell 15, 13th Gen Intel Core i5-1334U (16GB DD...",3.3,"39,990"
5,"Acer Smartchoice Aspire One, Intel Core Celero...",4.0,"50,490"
6,"Dell 15 (Previously Inspiron) Laptop, 14th Gen...",4.2,"72,120"
7,"acer Aspire 7, Intel Core 5-210H Processor, NV...",4.8,"49,480"
8,"HP 15, 13th Gen Intel Core i3-1315U Laptop (8G...",4.1,"68,990"
9,"ASUS TUF A15 (2025), AMD Ryzen 7 7445HS,RTX 30...",3.4,"63,990"


In [85]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Laptop   14 non-null     object
 1   Ratings  14 non-null     object
 2   prices   14 non-null     object
dtypes: object(3)
memory usage: 468.0+ bytes


In [86]:
data

,Laptop,Ratings,prices
0,"ASUS Vivobook 16,Smartchoice, 13th Gen, Intel ...",4.2,"63,990"
1,"HP 15, AMD Ryzen 5 7520U,(16GB LPDDR5, 512GB S...",4.0,"54,690"
2,Lenovo V15 G4 AMD Athlon Silver 7120U Laptop 8...,4.0,"43,999"
3,More results,3.8,"59,990"
4,"Dell 15, 13th Gen Intel Core i5-1334U (16GB DD...",3.3,"39,990"
5,"Acer Smartchoice Aspire One, Intel Core Celero...",4.0,"50,490"
6,"Dell 15 (Previously Inspiron) Laptop, 14th Gen...",4.2,"72,120"
7,"acer Aspire 7, Intel Core 5-210H Processor, NV...",4.8,"49,480"
8,"HP 15, 13th Gen Intel Core i3-1315U Laptop (8G...",4.1,"68,990"
9,"ASUS TUF A15 (2025), AMD Ryzen 7 7445HS,RTX 30...",3.4,"63,990"


In [91]:
data['prices'] = data['prices'].apply(lambda x: int(x.replace(',', '')))

In [92]:
data

,Laptop,Ratings,prices
0,"ASUS Vivobook 16,Smartchoice, 13th Gen, Intel ...",4.2,63990
1,"HP 15, AMD Ryzen 5 7520U,(16GB LPDDR5, 512GB S...",4.0,54690
2,Lenovo V15 G4 AMD Athlon Silver 7120U Laptop 8...,4.0,43999
3,More results,3.8,59990
4,"Dell 15, 13th Gen Intel Core i5-1334U (16GB DD...",3.3,39990
5,"Acer Smartchoice Aspire One, Intel Core Celero...",4.0,50490
6,"Dell 15 (Previously Inspiron) Laptop, 14th Gen...",4.2,72120
7,"acer Aspire 7, Intel Core 5-210H Processor, NV...",4.8,49480
8,"HP 15, 13th Gen Intel Core i3-1315U Laptop (8G...",4.1,68990
9,"ASUS TUF A15 (2025), AMD Ryzen 7 7445HS,RTX 30...",3.4,63990


In [ ]:
pd.read